feat(utxo-lib): add updateWalletOutputForPsbt

davidkaplanbitgo opened the pull request:
Refactor out output updating from adding in the same way as we do for inputs. This is useful for partial updating of wallet outputs for hydration.

TICKET: BTC-284

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/wallet/WalletOutput.ts:
```diff
@@ -29,14 +30,67 @@ export function addWalletOutputToPsbt(
+    );
+  }
+
+  // It is expensive to attempt to compute every output address using psbt.txOutputs[outputIndex]
+  // to then just get the script. Here, we are doing the same thing as what txOutputs() does in
+  // bitcoinjs-lib, but without iterating over each output.
+  const outputScript: Buffer = (psbt as any).__CACHE.__TX.outs[outputIndex].script;
```
This is kind of ugly, but it allows us to not call `const outputScript = psbt.txOutputs[outputIndex]` for each iteration. Additionally, we do not even try to make any addresses from the scripts, so it is very efficient

OttoAllmendinger replied:
let's add a method `getOutputScript(outputIndex): Buffer` to UtxoPsbt then (maybe we can upstream it)

davidkaplanbitgo commented:
Failing due to vulnerabilities, nothing wrong with the PR

davidkaplanbitgo commented:
Rebased off of https://github.com/BitGo/BitGoJS/pull/3729

davidkaplanbitgo committed:
test(utxo-lib): add specific outputs when creating psbt

BTC-284

davidkaplanbitgo committed:
feat(utxo-lib): efficiently get output script at index

BTC-284

davidkaplanbitgo committed:
feat(utxo-lib): add updateWalletOutputForPsbt

Refactor out output updating from adding in the same
way as we do for inputs on PSBTs. This is useful for
partial updating of wallet outputs for hydration.

TICKET: BTC-284
